prediction_market_agent/db/sql_handler.py

```diff
@@ -11,31 +14,41 @@
 
 class SQLHandler:
     def __init__(
-        self, model: t.Type[SQLModelType], sqlalchemy_db_url: str | None = None
+        self,
+        model: t.Type[SQLModelType],
+        sqlalchemy_db_url: str | None = None,
     ):
-        self.engine = create_engine(
+        self.sqlalchemy_db_url = (
             sqlalchemy_db_url
-            if sqlalchemy_db_url
-            else check_not_none(DBKeys().SQLALCHEMY_DB_URL)
+            if sqlalchemy_db_url is not None
+            else check_not_none(DBKeys().SQLALCHEMY_DB_URL).get_secret_value()
         )
+
+        api_keys = APIKeys_PMAT(SQLALCHEMY_DB_URL=SecretStr(self.sqlalchemy_db_url))
+        self.db_manager = DBManager(api_keys)
         self.table = model
         self._init_table_if_not_exists()
 
     def _init_table_if_not_exists(self) -> None:
-        table = SQLModel.metadata.tables[str(self.table.__tablename__)]
-        SQLModel.metadata.create_all(self.engine, tables=[table])
+        self.db_manager.create_tables(sqlmodel_tables=[self.table])
 
     def get_all(self) -> t.Sequence[SQLModelType]:
-        return Session(self.engine).query(self.table).all()
+        with self.db_manager.get_session() as session:
+            return session.query(self.table).all()
 
     def save_multiple(self, items: t.Sequence[SQLModelType]) -> None:
-        with Session(self.engine) as session:
+        with self.db_manager.get_session() as session:
             session.add_all(items)
             session.commit()
 
-    def delete_all_entries(self, col_name: str, col_value: str) -> None:
-        with Session(self.engine) as session:
-            session.query(self.table).filter_by(**{col_name: col_value}).delete()
+    def delete_all_entries(
+        self, col_name: str | None = None, col_value: str | None = None
```

Comment:
I guess this is the glitch that allowed to delete the whole table? Any reason to have it here, instead of conditioned as before? 

I don't see a use-case where some agent should delete the whole table, instead of just based on some condition, do you?

Reply:
Ah I see now, tests, because dbmanager is session scoped.

What about having `truncate` utility function in tests and calling it there?

Reply:
Based on SQLModel [docs](https://sqlmodel.tiangolo.com/tutorial/delete/#commit-the-session), the way to go is actually to call `Session(engine).delete()` and then `commit()`.
The accident came from calling `SQLModel.metadata.drop_all` (on the wrong instance of the DB), not from this method. Hence I would argue that this logic is OK as is.

Reply:
It still leaves space to just accidentally delete everything, while it's only used in tests (which can be with almost zero effort achieved by having `delete_all_entries` defined as a utility function in tests).

However, if you believe that either we or some agent using this function potentially in the future will never make such a mistake, okay by me, I'll just use this meme if it happens in the end 😄 

![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/e9f3f2a9-1b1b-48b0-9c01-c9dfb70bda1a)

Reply:
I'm actually also thinking about whether using `delete_all_entries` in the tests is good practice, I'd say no.

As that also leaves space to just accidentally drop everything. Wouldn't it be a better way to figure out a way how to do it without it? (DBManager lives within sessions, that's clear, but maybe we could somehow remove its instances to drop the old connection, instead of manually deleting all entries).

Btw in PMAT, no `delete_all_entries` wasn't necessary after switching to a session-scoped database connection,  but I guess these tests were failing without it?

Reply:
My suggestion would be:
-> Removing `delete_all_entries from DBManager` to avoid errors in code logic
-> Making the db_manager `function specific` ([scope function](https://docs.pytest.org/en/stable/how-to/fixtures.html#fixture-scopes)) so that the DB data is automatically rollbacked when tests finish

Would that address all your concerns here, or anything left to discuss?

Reply:
Yes that sounds good!

Reply:
Implemented, see https://github.com/gnosis/prediction-market-agent/pull/588/files#diff-168e93f323de4635a6e1182e9d23e6dbdbb1fdcdbd45ec3cb9d3265e98d91214R22-R25